pallets/prices/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+        type StableCurrencyFixedPrice: Get<OraclePrice>;
+
+        /// The origin which may set prices feed to system.
+        type FeederOrigin: EnsureOrigin<Self::Origin>;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Set emergency price. \[currency_id, price_detail\]
+        SetPrice(CurrencyId, OraclePrice),
+        /// Reset emergency price. \[currency_id\]
+        ResetPrice(CurrencyId),
+    }
+
+    /// Mapping from currency id to it's emergency price
+    #[pallet::storage]
+    #[pallet::getter(fn emergency_price)]
+    pub type EmergencyPrice<T: Config> =
+        StorageMap<_, Twox64Concat, CurrencyId, OraclePrice, OptionQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Set emergency price
+        #[pallet::weight(100)]
+        #[transactional]
+        pub fn set_price(
+            origin: OriginFor<T>,
+            currency_id: CurrencyId,
+            price: OraclePrice,
+        ) -> DispatchResultWithPostInfo {
+            T::FeederOrigin::ensure_origin(origin)?;
+            <Pallet<T> as EmergencyPriceFeeder<CurrencyId, OraclePrice>>::set_emergency_price(
+                currency_id,
+                price,
+            );
+            Ok(().into())
+        }
+
+        /// Reset emergency price
+        #[pallet::weight(100)]
+        #[transactional]
+        pub fn reset_price(
+            origin: OriginFor<T>,
+            currency_id: CurrencyId,
+        ) -> DispatchResultWithPostInfo {
+            T::FeederOrigin::ensure_origin(origin)?;
+            <Pallet<T> as EmergencyPriceFeeder<CurrencyId, OraclePrice>>::reset_emergency_price(
+                currency_id,
+            );
+            Ok(().into())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    // get emergency price, the timestamp is zero
+    fn get_emergency_price(currency_id: &CurrencyId) -> Option<PriceDetail> {
+        if let Some(price) = Self::emergency_price(currency_id) {
+            Some((price.into_inner(), 0))
+        } else {
+            None
+        }
+    }
+}
+
+impl<T: Config> PriceFeeder for Pallet<T> {
+    /// Get price and timestamp by currency_id
+    /// Timestamp is zero and the currency is not stable currency means the price is emergency price
+    fn get_price(currency_id: &CurrencyId) -> Option<PriceDetail> {
+        // if is stable currency, return fixed price
+        if *currency_id == T::GetStableCurrencyId::get() {
+            Some((T::StableCurrencyFixedPrice::get().into_inner(), 0))
+        } else {
+            // if emergency price exists, return it, otherwise return latest price from oracle.
+            Self::get_emergency_price(currency_id).or_else(|| {
+                T::Source::get_no_op(&currency_id)
```

Comment:
as I check, `get_no_op` won't store the combined data, so we need to combine t every time, I am not sure if this is ok?

Reply:
Fixed